src/Integration/MefServices/ProjectsRuleSetProvider.cs

```diff
@@ -0,0 +1,355 @@
+                if (projectsWithSonarRuleSet.Count > 0)
+                {
+                    this.logger.WriteLine(Strings.WarningProjectsWithSonarRulesWhileNewConnected,
+                        string.Join(", ", projectsWithSonarRuleSet));
+                }
+            }
+        }
+
+        private async void OnUpdatedOnDisk(object sender, EventArgs e)
+        {
+            var fileChangeTracker = sender as FileChangeTracker;
+            if (fileChangeTracker == null)
+            {
+                return;
+            }
+
+            if (!fileChangeTracker.FilePath.EndsWith(".ruleset"))
+            {
+                var cachedProjectRuleSetPath = projectPathToCachedData[fileChangeTracker.FilePath].RuleSetPath;
+
+                var project = projectSystemHelper.GetSolutionProjects().FirstOrDefault(p =>
+                    p.FullName.Equals(fileChangeTracker.FilePath, StringComparison.OrdinalIgnoreCase));
+                if (project != null)
+                {
+                    var projectRuleSetPath = projectSystemHelper.GetProjectProperty(project, Constants.CodeAnalysisRuleSetPropertyKey);
+                    var projectDirectoryFullPath = new FileInfo(project.FullName).Directory.FullName;
+                    var projectRuleSetFullPath = GetFullPath(projectRuleSetPath, projectDirectoryFullPath);
+
+                    if (cachedProjectRuleSetPath.Equals(projectRuleSetFullPath, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Project changed but the ruleset property is the same so there is no need to recreate the cache
+                        return;
+                    }
+                }
+                else
+                {
+                    // Cannot find the project (most likely project was changed while being unloaded) so let's rebuild the cache
+                    // in case the CodeAnalysisRuleSet property was changed
+                }
+            }
+
+            ClearCache();
+            await BuildCacheAsync();
+        }
+
+        /// <remarks>
+        /// [SIDE EFFECT] The set of ruleset is update at each call.
+        /// </summaryremarks>
+        private ProjectData ProcessProjectAndCollectFilesToTrack(EnvDTE.Project project)
+        {
+            TrackIfNotTracked(project.FullName);
+
+            var projectRuleSetPath = projectSystemHelper.GetProjectProperty(project, Constants.CodeAnalysisRuleSetPropertyKey);
+            var projectDirectoryFullPath = new FileInfo(project.FullName).Directory.FullName;
+            var projectRuleSetFullPath = GetFullPath(projectRuleSetPath, projectDirectoryFullPath);
+
+            var hasAnySonarRule = false;
+
+            if (File.Exists(projectRuleSetFullPath))
+            {
+                var projectRuleSet = RuleSet.LoadFromFile(projectRuleSetFullPath);
+
+                // 1. Collect all paths (current ruleset + includes)
+                var ruleSetIncludeFullPaths = projectRuleSet.RuleSetIncludes
```

Comment:
An included ruleset can include other rulesets, can't it?

Reply:
Good point. That might be tricky to handle though...